Add TrafficControl API

tnqn opened the pull request:
TrafficControl is a feature which allows mirroring or redirecting the
traffic Pods send or receive. It enables users to monitor and analyze
Pod traffic, and to enforce custom network protections for Pods with
fine-grained control over network traffic.

This patch adds types and CRD for TrafficControl API.

Examples:

1. Mirror Pods (web=app) ingress traffic to a VXLAN tunnel
```
apiVersion: crd.antrea.io/v1alpha2
kind: TrafficControl
metadata:
  name: mirror-web-app
spec:
  appliedTo:
    podSelector:
      matchLabels:
        app: web
  direction: Ingress
  action: Mirror
  targetPort:
    vxlan:
      remoteIP: 1.1.1.1
```

2. Redirect Pods (web=app) traffic in both direction to OVS internal
port firewall0 and expect the traffic to re-enter OVS via another OVS
internal port firewall1 if they are not dropped.
```
apiVersion: crd.antrea.io/v1alpha2
kind: TrafficControl
metadata:
  name: redirect
spec:
  appliedTo:
    podSelector:
      matchLabels:
        role: web
  direction: Ingress
  action: Redirect
  targetPort:
    ovsInternal:
      name: firewall0
  returnPort:
    ovsInternal:
      name: firewall1
```

For #3324

Signed-off-by: Quan Tian <qtian@vmware.com>

jianjuns commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,116 @@ type IPPoolList struct {
+	Port *PortDevice `json:"port,omitempty"`
+	// VXLAN represents a VXLAN tunnel that is created on the Node.
+	// +optional
+	VXLAN *TunnelDevice `json:"vxlan,omitempty"`
```
Should we just add a type to TunnelDevice?

tnqn replied:
ERSPAN is kind of Tunnel, if we add a type to TunnelPort, ERSPAN struct would repeat some tunnel attributes.

jianjuns replied:
I feel that is ok. Or if we follow ovs-vsctl, we can put ERSPAN parameters into TunnelPort

jianjuns replied:
After more thinking, how about:

OVSPort struct {
    Name
    Type // Internal, Device, GRE, GENEVE, VXLAN, ERSPAN
    Tunnel // Tunnel parameters (remote IP, etc.)
    ERSPAN // ERSPAN parameters
}

tnqn replied:
If I get you correctly, it uses a type field to indicate type and have all type specific parameters in single struct. I feel ovs-vsctl organizes parameters in this way because it needs a consistent db schema for all kinds (all fields are in same level in form of key value). It seems opposite to how K8s API organizes type specific parameters.

And I feel organizing type specific parameters in the type's own struct is more clear and easier to validate and use. For users, it's clear to see which parameters a kind of port has, no need to understand the meaning of irrelevant fields. For developers, it's easier to consume the type specific fields when handling one type. Otherwise we need comments to explain which fields are needed for which types for both users and developers.

Examples of K8s APIs: 
https://github.com/kubernetes/kubernetes/blob/1108bed7631f545d43530aa697175d243b99610b/pkg/apis/core/types.go#L58
https://github.com/kubernetes/kubernetes/blob/1108bed7631f545d43530aa697175d243b99610b/pkg/apis/core/types.go#L2234

jianjuns replied:
Sure. Your new version looks good to me too. I think it is just about sharing and encapsulation.

jianjuns replied:
The only thing is LocalPort looks strange, as it is just a flag and it is needed only by internal port.

jianjuns commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,116 @@ type IPPoolList struct {
+// configuration.
+type Device struct {
+	// The name of the device.
+	Name string `json:"returnDevice"`
```
returnDevice -> name

tnqn replied:
fixed

tnqn commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,116 @@ type IPPoolList struct {
+// return device of traffic redirecting.
+// Multiple TrafficControl objects can use same device, however, the devices with same name must have consistent
+// configuration.
+type Device struct {
```
I wanted to use the parent type name as the suffix of the child type. But you proposal makes sense, we don't have to add it.

jianjuns commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,123 @@ type IPPoolList struct {
+
+// LocalPort represents a local port that is attached to the OVS bridge. It can be an OVS internal port, a physical NIC,
+// or a veth device.
+type LocalPort struct {
```
This one looks strange, as it is shared by two types of ports, and just a flag to indicate which type. From OVS perspective, internal port is a separate type from port with a device attachment; they are not closer than tunnel type port? This is the major reason that I suggested to move type out to TrafficControlPort.

But if you want to go this way, how about adding a `Device` field to LocalPort. The port name can device name can be different, and also LocalPort is more meaningful then.

tnqn replied:
Creating a port name different from device name doesn't seem useful, and it would be a little redundant for most cases in which port name and device name are same. And it's still mixing type specific parameters in one struct when there are already type specific sub-structs.

Another way I'm thinking is similar to what you suggested in https://github.com/antrea-io/antrea/pull/3644#discussion_r850942266:
```
type TrafficControlPort struct {
	// OVSInternal represents an OVS internal port that is created on the Node.
	// +optional
	OVSInternal *OVSInternalPort `json:"ovsInternal"`
	// Device represents a network device that is attached to the OVS bridge. It can be a physical NIC or a veth device.
	// +optional
	Device *DevicePort `json:"Device,omitempty"`
	// VXLAN represents a VXLAN tunnel that is created on the Node.
	// +optional
	VXLAN *TunnelPort `json:"vxlan,omitempty"`
	// GENEVE represents a GENEVE tunnel that is created on the Node.
	// +optional
	GENEVE *TunnelPort `json:"geneve,omitempty"`
	// GRE represents a GRE tunnel that is created on the Node.
	// +optional
	GRE *TunnelPort `json:"gre,omitempty"`
	// ERSPAN represents an ERSPAN tunnel that is created on the Node.
	// +optional
	ERSPAN *ERSPANPort `json:"erspan,omitempty"`
}

type OVSInternalPort struct {
	Name string `json:"name"`
}

type DevicePort struct {
	Name string `json:"name"`
}

// TunnelPort represents a tunnel that is created on the Node.
type TunnelPort struct {
	// The remote IP of the tunnel.
	RemoteIP string `json:"remoteIP"`
	// The ID of the tunnel.
	TunnelID int64 `json:"tunnelID,omitempty"`
}
```
And we don't have Name field in the parent struct. The port name of tunnel will be generated based on the tunnel's attributes so there wouldn't be conflict in theory. If tunnel's attributes are same, they will use same port, otherwise they will use different ports. For internal port and device port, if users set same name in different tc objects, it already means they want to use same port.

jianjuns replied:
I still prefer:

```
OVSPort struct {
    Name
    Type // Internal, Device, GRE, GENEVE, VXLAN, ERSPAN
    Tunnel // Tunnel parameters (remote IP, etc.)
    ERSPAN // ERSPAN parameters
}
```

It is easiest to understand. If for the API validation reason, you think should not have the Type field, your proposal is fine to me too. I think we can still add Name to TunnelPort assuming then the implementation can be simpler.

tnqn replied:
I'm ok with moving tunnel type  to TunnelPort if you prefer. Just want to show how OVSInternalPort and DevicePort will be differentiated here. The reasons why moving Name to `OVSInternalPort` and `DevicePort` are:
1. Otherwise `OVSInternalPort` and `DevicePort` would have no attributes, which looks a little weird when setting this type as it would be
```
targetPort:
  name: port0
  ovsInternal: {}
```
2. As you mentioned, user don't really care tunnel port name even though they exist, but user care ovsInternalPort name device name as they will need to listen on the ports.
3. Avoid conflict when ports have same name but are of different types.

tnqn replied:
Note that even if we extract TrafficControlPort to a separate API later, we cannot just use port name as the object name as they have different format requirement. Object name can be longer than port name. So the object name can only be used as a pointer from TrafficControl object to TrafficControlPort object.

jianjuns replied:
> Otherwise OVSInternalPort and DevicePort would have no attributes, which looks a little weird

That is why I feel adding Type to the top level struct makes sense. I did not check, but I assume there must be many such cases in K8s API, like Service type.

> Avoid conflict when ports have same name but are of different types.

Removing Name from Tunnel ports does not solve the problem, as Internal and Device port can conflict too, and in theory users can create an Internal port with a tunnel port name. I meant it seems a simpler assumption, we require no port sharing between TrafficControl CRs in this version. And later if we support creating port through a different mechanism or pre-creating manually, then it is consistent port has a user specified name. But I do not have a strong opinion on this one.

tnqn replied:
> I still prefer:
> 
> ```
> OVSPort struct {
>     Name
>     Type // Internal, Device, GRE, GENEVE, VXLAN, ERSPAN
>     Tunnel // Tunnel parameters (remote IP, etc.)
>     ERSPAN // ERSPAN parameters
> }
> ```
> 
> It is easiest to understand. If for the API validation reason, you think should not have the Type field, your proposal is fine to me too. I think we can still add Name to TunnelPort assuming then the implementation can be simpler.

The problem of this way is we need to tell users and developers:
For Internal port, you should set Name and Type fields.
For Device port, you should set Name and Type fields.
For GRE, GENEVE, VXLAN ports, you should set Name, Type, and Tunnel fields.
For ERSPAN port, you should set all fields.

It may be more complicated if we support more types like RSPAN etc.

jianjuns replied:
> So the object name can only be used as a pointer from TrafficControl object to TrafficControlPort object.

I am not sure. Why it cannot be just port name. Port can be created by user manually or another solution. I am not sure we must have a CRD to create a single port either (e.g. it can be ConfigMap, etc.).

jianjuns replied:
> The problem of this way is we need to tell users and developers:

Why it is a problem? Why you should not set tunnel attributes to create a tunnel? If we say every sub-type must be self contained, then there is no sharing and inheritance/extension.

tnqn replied:
> > So the object name can only be used as a pointer from TrafficControl object to TrafficControlPort object.
> 
> I am not sure. Why it cannot be just port name. Port can be created by user manually or another solution. I am not sure we must have a CRD to create a single port either (e.g. it can be ConfigMap, etc.).

I mean if we extract TrafficControlPort to a separate CRD. I feel using a ConfigMap whose data is less structured and lack of validation is not better than using a CRD, especially when they are referred by another CRD. For other solutions or created by user manually, since the OVS resources are not available in default path, I think the user experience wouldn't be good to ask user to install ovs resources and specify antrea specific path to operate OVS, or mount antrea volumes in another daemonset to operate OVS. If I was told to do such operations for troubleshooting purpose, I would still install tcpdump or run a daemonset pod with tcpdump tool directly.

jianjuns replied:
I think we do know people wanted to do such integration earlier (configuring port, etc. by their DaemonSet)? Of course It is not application developers, but some network solutions. But I think this is a minor problem. I feel requiring a Port name can make the API looks uniform and set the shared assumption Port cannot be shared. It might help people with less knowledge troubleshooting too. And it is not a big overhead for users. So, I feel no reason to remove it just for making one possible API design looks better. But again I do not have a strong opinion on this one, and wont insist.

I feel we should just decide whether a Type field makes sense, or really not a convention in K8s API.

tnqn replied:
> I think we do know people wanted to do such integration earlier (configuring port, etc. by their DaemonSet)?

I'm not sure if they wanted to or had to when antrea doesn't provide the capacity. It's like providing API for interaction vs. letting another component interact with antrea's internal directly. I consider ovs resources as antrea internal because it manages ovs's lifecycle, uses custom path that is specific to antrea, configures ovs exclusively (drops traffic from/to any unknown port). I think we don't really want other solutions make any assumption on how OVS is used in antrea, for example, if another solution adds a flow to overwrite some antrea behavior by assuming priority 210 is the largest priority antrea uses or assuming table 70 is the L3forwarding table, it would stop working when upgrading to v1.6.0.

>> Avoid conflict when ports have same name but are of different types.

> Removing Name from Tunnel ports does not solve the problem, as Internal and Device port can conflict too, and in theory users can create an Internal port with a tunnel port name. I meant it seems a simpler assumption, we require no port sharing between TrafficControl CRs in this version. And later if we support creating port through a different mechanism or pre-creating manually, then it is consistent port has a user specified name. But I do not have a strong opinion on this one.

> I feel requiring a Port name can make the API looks uniform and set the shared assumption Port cannot be shared. It might help people with less knowledge troubleshooting too. And it is not a big overhead for users.

Sorry, I don't quite get it. Let me rephrase my points:
For internal port and device port, "Name" is the attribute uniquely identifying a destination. If there is an internal port named "port0", and user create another device port named "port0", it's misconfiguration, not a case we can or need to support as user already know "port0" is not a device (if user can receive traffic from devices, they have access to all local devices), and this misconfiguration can be easily reported via realization status or event and corrected by user. If there is an internal port named "port0", and user create another tc using "port0" with same type, this literally reuses the port and there seems to be nothing wrong with it.

For tunnel ports, "TunnelID" + "RemoteIP" are the attributes uniquely identifying a destination. 
* If we provide a configurable port name to tunnel types, we make conflict possible. If "Name" is same but any of "TunnelID" and "RemoteIP" are different, one would overwrite another. If "Name" is different but both "TunnelID" and "RemoteIP" are same, it doesn't work either as creating tunnel port with same config would fail immediately. And the "Name" is actually meaningless to user: there is no corresponding network device user can see or debug. Even if we report the two conflicts to user, the error would look confusing and they can do nothing about them:
  * If it's because of same "Name" and different "TunnelID+RemoteIP", it's hard for users to know all existing OVS port names and avoid conflict if they don't have access to antrea pod. And when we create such tunnels, many people would just start from "vxlan0", "vxlan1", and is hard to explain why they need to resolve the name conflict when they just want to mirror the traffic to their own tunnel endpoints.
  * If it's because of different "Name" and same "TunnelID+RemoteIP", the tunnel endpoint is likely managed by same user but they need to figure out the Name they used in other TC objects and make them consistent. Still, what's the point of introducing a meaningless field that can cause conflict.
* If we do not provide a configurable port name to tunnel types, and generate the port name based on "TunnelID+RemoteIP", we make conflict impossible. If "TunnelID+RemoteIP" are same, it literally reuses the tunnel and nothing wrong with it. If "TunnelID+RemoteIP" are different, they have different port names.

I think sharing port or tunnel across multiple TC objects can be easily supported with above implementation, we don't have to add this limitation, which is not very friendly and reasonable.

> I feel we should just decide whether a Type field makes sense, or really not a convention in K8s API.

I would not say having a Type field is not a convention in K8s API. It's just which style we use to indicate Type, a separate field, or the key of the sub-struct. I believe there are many cases for both in K8s and Antrea API. Even for TC object, we have an optional field "ReturnDevice" that is specific to Redirect action, it can be organized in another way that uses "Mirror", "Redirect" as keys but it's obviously overkill to have two sub-structs just because of an optional field. The cases that use Type as key and have sub-structs for each type normally have a number of type specific parameters, which makes structs look cleaner, e.g. various protocols, various probes, various volumes.

But whether it really looks cleaner is subjective so we didn't reach agreement on it. I wouldn't insist on this if you prefer adding a Type field. But I want to list some potential "issues" for reference:
1. I don't know if you mean having `inline` tags for the sub-structs, but it would affect how user set the parameters. * If we add `inline`, then all parameters from sub-structs will be in same level, fields with generic names like `Version` and `Index` in ERSPAN in the first level would look confusing and be easy to conflict when extending the support for other types and perhaps need a prefix "ERSPAN" to these fields, even when they are already in a struct called "ERSPAN".
* If we do not add `inline`, then user will need to set multiple sets of parameters to configure cases like ERSPAN:
```
type: ERSPAN
tunnel:
  remoteIP: x.x.x.x
  tunnelID: 1
erspan:
  version: xx
  index: xx
```

2. If the above points about not providing configurable Port Name for tunnels make sense, it would be more hard for users to get which fields are required for which types when generic fields like "Name", "Version" exist in same struct but should be only set for certain types.

And for the way that uses type as key, I wanted to avoid having empty structs for certain types but I don't feel it's a problem even if some sub-structs are empty, we just need to document the type of the tunnel should be the key, and parameters of the type should be the values. Moving Name field to OVSInternalPort and DevicePort helps avoid empty structs but it's not just for it. 

BTW, if we want to distinguish OVSInternalPort and DevicePort, Name actually has different means, one is for the name of the internal port we will create, another one is for the name of the device we will attach to OVS. If we don't want to distinguish OVSInternalPort and DevicePort, I grouped them in LocalPort because both of them are local to the Node and user can receive traffic of an interface with that name from usage perspective, I think we don't need to care whether it's internal, pNIC, or veth. Or perhaps I should use `AutoCreate` for LocalPort to represent whether user want antrea to create the port or they want to use an existing port and do not make it specific to device type.

jianjuns replied:
For manual port creation or by another solution:
> I'm not sure if they wanted to or had to when antrea doesn't provide the capacity. 

I meant it is possible Antrea does not support all configuration or wants not to support all configuration which users or other solutions (e.g. some NSX-T features) want. I saw this kind of "low-level" integration happens in real world.

I understood your point that port name is not a must and can be another potential source of conflicts. But I do not get why it is a big deal. Many your points like users might fail to track port name, etc. apply to internal port as well (if we really want to avoid the conflict, we can auto-generate a name for internal ports, and ask consumer to use the auto-generated name). You assume there can be multiple users create TC/ports. I do not think so. In reality only a few ports will be created, and the creator should know all them (esp. for one tunnel destination).

When you say multiple TC's share a tunnel/port. Do you mean this version? Or later when we have a separate mechanism to create port? If you mean the first version, besides implementation complexity, I do not like users to duplicate the port/tunnel/ERSPAN configuration in multiple TC's. It is another source to introduce misconfiguration. BTW, when you have a separate CR to create port, the CR name has the same problems as port name as you described.

What you think about my point of API uniformity? I meant it helps understanding if all sub-options of ports share a common mandatory name (which might be case for people has little knowledge on OVS and different port types/tunnels; and also people know OVS well and know these are indeed realized with an OVS port which has a unique name). I do not really want to hide OVS layer either. I hope more users can understand OVS and do some basic troubleshooting. Exposing port name is good for this purpose.

jianjuns replied:
In summary, you and I proposed two choices:

1)
```
OVSPort struct {
    Name string
    Type PortType // Internal, Device, GRE, GENEVE, VXLAN, ERSPAN
    TunnelConfig TunnelConfig // Tunnel parameters (remote IP, etc.)
    ERSPANConfig ERSPANConfig // ERSPAN parameters. **I would duplicate tunnel parameters in ERSPANConfig.**
}
```

2)
```
type TrafficControlPort struct {
	OVSInternal *OVSInternalPort `json:"ovsInternal,omitempty"`
	Device *DevicePort `json:"device,omitempty"`
	VXLAN *TunnelPort `json:"vxlan,omitempty"`
	GENEVE *TunnelPort `json:"geneve,omitempty"`
	GRE *TunnelPort `json:"gre,omitempty"`
	ERSPAN *ERSPANPort `json:"erspan,omitempty"`
}

type OVSInternalPort struct {
        // The name of the OVS internal port.
	Name string `json:"name"`
}

type DevicePort struct {
	Name string `json:"name"`
}

type TunnelPort struct {
	RemoteIP string `json:"remoteIP"`
	TunnelID int64 `json:"tunnelID,omitempty"`
}
```

jianjuns replied:
I still prefer 1). 2) is ok to me too, but I like to consider at least an optional port name in TunnelPort.

Let alone all our discussion, for people like me who know about OVS, 1) looks natural; and even we go 2) I may like to control the tunnel port name. And I feel users/solutions consume the feature, do know about OVS.

tnqn replied:
> What you think about my point of API uniformity? I meant it helps understanding if all sub-options of ports share a common mandatory name (which might be case for people has little knowledge on OVS and different port types/tunnels; and also people know OVS well and know these are indeed realized with an OVS port which has a unique name). I do not really want to hide OVS layer either. I hope more users can understand OVS and do some basic troubleshooting. Exposing port name is good for this purpose.

If we consider this feature only for cluster admin/provider, yes, they have access to OVS resources and may have the knowledge. I changed the design when considering whether this feature can be used for the cluster's user, who cannot access antrea pods, cannot access host network, and the cluster's PodSecurityPolicy doesn't allow users to create priviledeged or hostNetwork Pod (which may be common for production clusters?), if we can make it possible to mirror their own Pods' traffic to their own tunnel endpoints for monitoring or troubleshooting, would it be a cool feature? For such users, avoiding port name conflict is a hard requirement but it's easy to happen. Of course, this needs more fine-grained permission control on TC object itself in the future, but making tunnel port name not mandatory is a basic requirement for that. Let me know if you think providing traffic mirroring (to tunnel) capacity to normal users is a reasonable consideration.

> BTW, when you have a separate CR to create port, the CR name has the same problems as port name as you described.

If we have a separate CR to create port, the CR name cannot be used as port names anyway as CR name can be longer than port name, and their allowed charactors may be different. The CR name will only be used as a pointer from TC object to Port object. Like:
```
Port:
  name: my-network-monitor
  spec:
    vxlan:
      remoteIP: 1.1.1.1
---
TrafficControl:
  name: mirror-web-app
  spec:
    appliedTo:
      app: web
    action: mirror
    targetPort: my-network-monitor
---
TrafficControl:
  name: mirror-backend-app
  spec:
    appliedTo:
      app: backend
    action: mirror
    targetPort: my-network-monitor
```

jianjuns replied:
Again, I do not feel port name is a big deal. Even the use case you described are by users know about networking very well (otherwise they wont set up tunnel on the consumer side and consume raw packets), and why there can be many such tunnels by multiple users? But even you want to consider that, we can make port name optional (I feel it is not worthwhile and would make it mandatory for now, but up to you).

I do feel we should over-engineer to do fine-grained control to allow all users to mirror particular traffic. I never see the requirement for low level switching features. Even there is such a use case, they can request cluster admin to do that.

For a separate CR, I meant CR name must be unique too. But I guess your point is CR name has no length limit and so easier to generate unique name? It is true, but still users need to understand two CRs cannot share the same parameters as they end up to be a single port on OVS.

tnqn replied:
> Again, I do not feel port name is a big deal. Even the use case you described are by users know about networking very well (otherwise they wont set up tunnel on the consumer side and consume raw packets), and why there can be many such tunnels by multiple users? 

We should first say whether this can be used by tenant users in the future. If yes, and port name is mandatory but its naming scope is global, users in Namespace A can create vxlan0 pointing to their endpoints, then users in Namespace B cannot use vxlan0 as port name and they will have no idea why is it because they don't have global visibility. If this is tenant user facing future, why we don't consider multiple tunnels by multiple users?

> But even you want to consider that, we can make port name optional (I feel it is not worthwhile and would make it mandatory for now, but up to you).

If the port name is optional for tunnels, then 1) is less suboptimal than 2) in my mind as API uniformity is not met anyway and there would be more docs explaining some fields must be set for some types and cannot set for other types, some fields are mandatory for some types but optional for other types, more complicated self-contained structs.

> I do feel we should over-engineer to do fine-grained control to allow all users to mirror particular traffic. I never see the requirement for low level switching features. Even there is such a use case, they can request cluster admin to do that.

Guess you wanted to say "I do NOT feel". OK, without the use case, I would just set Name as mandatory and go 1).

> For a separate CR, I meant CR name must be unique too. But I guess your point is CR name has no length limit and so easier to generate unique name? It is true, but still users need to understand two CRs cannot share the same parameters as they end up to be a single port on OVS.

No, I meant for users who have global visibility, they know all Names and it's their responsibility to avoid CR name conflict, just like any of other cluster scoped resources. For tenant users, they don't know CR names / Port names in other Namespaces, it doesn't make sense to require them to resolve name conflict with CR or Ports declared in other Namespaces. Again, this is only for future extensibilty if we want to provide fine-grained (namespaced) control to tenant users. If this is not a requirement, no need to discuss this further.

jianjuns replied:
I do not think "tenant" user should directly create tunnels to redirect traffic to remote (and even if that happens we should define new API for that. Definitely not the current API which is for cluster admin).

jianjuns replied:
Humm... after more thinking, I feel what you said about tunnel port name makes great sense for users only want to mirror packets to a remote tunnel and do not use/know about internal and device port at all. My problem is I always thought about all the use cases together, and so felt users would like API uniformity. To be most flexible, seems we should make tunnel port name optional to cover both users care and do not care about it. But for simplicity, I guess you like to choose only one way? Have you received the tunnel/ERSPAN only requirement?

Another my problem is I want to keep the possibility of supporting pre-created ports, so in case we need some special port configuration, we need not to change Antrea upstream. When the port is pre-created, actually Antrea does not care it is port type but just needs a port name, and the sub-type structs way might not work well?

In general, as I do not see a good way to satisfy all cases, I feel we can start from a port type and a mandatory name (even it might be strange for tunnel only users, but it should not be a big deal for them either). Later, if we really want, we can make port name optional (for internal, tunnel, ERSPAN, and for device we can also choose a separate device name field to be clearer).

tnqn replied:
> Humm... after more thinking, I feel what you said about tunnel port name makes great sense for users only want to mirror packets to a remote tunnel and do not use/know about internal and device port at all. My problem is I always thought about all the use cases together, and so felt users would like API uniformity. To be most flexible, seems we should make tunnel port name optional to cover both users care and do not care about it. But for simplicity, I guess you like to choose only one way? Have you received the tunnel/ERSPAN only requirement?

Glad to know we reach a consensus on tunnel port name. Yes, I would prefer only one way if specifying tunnel port name doesn't give much value other than API uniformity (it should be not hard for users who have OVS troubleshooting ability to figure out which port is auto-generated for their tunnels by looking for the tunnel ip and id via ovs-vsctl). 

https://github.com/antrea-io/antrea/issues/3008 is tunnel only requirement. For traffic mirroring, I imagine mirroring to tunnel would be a more common use case, especially for users who want traffic visibility, monitoring, analyzing services via a centralized endpoint, instead of running multiple services on every Node (for listening internal/device ports) which may need another result aggregation process. Running centralized monitoring service remotely may make it easier to have something like monitoring/IDS as a service, especially when monitored Pods are running on multiple Nodes, such service's implementation only needs to start a network analyzer (can even be a Pod) and create a traffic mirroring object pointing to its IP.

Besides, from my investigation on similar features, most of them only support mirroring to remote:
[Configure Port Mirroring on VMware Cloud on AWS](https://docs.vmware.com/en/VMware-NSX-T-Data-Center/3.2/administration/GUID-E944FB55-0285-466C-8603-CD701B17BB20.html)
[Configure Port Mirroring on VMware NSX-T Data Center](https://docs.vmware.com/en/VMware-Cloud-on-AWS/services/com.vmware.vmc-aws-networking-security/GUID-3268A0D3-89D0-406F-B44F-156DD1A30E00.html)
[Cloud IDS in Google Cloud](https://cloud.google.com/intrusion-detection-system/docs/overview)
I haven't found a similar feature in K8s ecosystem, thought Antrea might be the first one providing such capacity for end users. 
  
> Another my problem is I want to keep the possibility of supporting pre-created ports, so in case we need some special port configuration, we need not to change Antrea upstream. When the port is pre-created, actually Antrea does not care it is port type but just needs a port name, and the sub-type structs way might not work well?

I feel making port type optional may be even more hard to explain and away from API uniformity? Perhaps the use case can be covered by a type like "Opaque"? It can work in either way:
```
targetPort:
  type: Opaque
  name: xxx
---
targetPort:
  opaque:
    name: xxx
```

> In general, as I do not see a good way to satisfy all cases, I feel we can start from a port type and a mandatory name (even it might be strange for tunnel only users, but it should not be a big deal for them either). Later, if we really want, we can make port name optional (for internal, tunnel, ERSPAN, and for device we can also choose a separate device name field to be clearer).

Sure, I have no problem with starting from the current version (a port type and a mandatory name). But I have one question before merging it:
Since traffic mirroring to tunnel requires almost no priviledge and can potentially work for tenant users, does it make sense to separate traffic mirroring and redirecting APIs in the begining. Then the API name could be more specific like above solutions, and it's possible to have a namespaced traffic mirroring API in the future. Traffic redirecting is unlikely to need that as it requires too many priviledges to work so only suitable for cluster admin/provider.

jianjuns replied:
You are right. If we believe mirroring to a tunnel is an important feature (but I do not believe it should be for a tenant though), then we better to create a separate API for that, then we do not have all these problems. But have you received any ask for mirroring with tunnel?

I feel the current approach is just trying to share more implementation and support tunnel with low cost. It is not a bad strategy either if the cost is really low (?).

jianjuns commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,110 @@ type IPPoolList struct {
+)
+
+// OVSPort represents an OVS port that can be used as the target of traffic mirroring or redirecting, and the return
+// port of traffic redirecting. Multiple TrafficControl objects can use same port, however, the ports with same name
```
Do we plan to support: "Multiple TrafficControl objects can use same port" in the current version?

tnqn replied:
I think there is no harm to support it with the statement that ports with same name must have same configuration:
If we don't allow it, it requires more code to validate the port is not duplicate with any of existing ports, or how we say it's not allowed?
If we allow it with the precondition, port sharing can work if it's configured correctly.

jianjuns replied:
We need to validate the configuration is the same, and we need to track the reference to know when to delete the port.

jianjuns replied:
Ok, we had long discussion on this and I think we are on the same page. How about you choose one based on complexity and your preference. If we want to remove port name for tunnel, then we should go your way of:

```
type TrafficControlPort struct {
	OVSInternal *OVSInternalPort `json:"ovsInternal,omitempty"`
	Device *DevicePort `json:"device,omitempty"`
	VXLAN *TunnelPort `json:"vxlan,omitempty"`
	GENEVE *TunnelPort `json:"geneve,omitempty"`
	GRE *TunnelPort `json:"gre,omitempty"`
	ERSPAN *ERSPANPort `json:"erspan,omitempty"`
}

// Maybe we just need a string named "InternalPort" in TrafficControlPort?
type InternalPort struct {
        // The name of the OVS internal port.
	Name string `json:"name"`
}

// How about NetworkDevice?
type DevicePort struct {
	Name string `json:"name"`
}

// Probably do not call it TunnelPort then, but just Tunnel?
type TunnelPort struct {
	RemoteIP string `json:"remoteIP"`
	TunnelID int64 `json:"tunnelID,omitempty"`
}
```

Again, my major problem is then not straightforward to support pre-created port, but maybe we can add a filed like PrecreatedPort/ExistingPort, etc. for that when we need to support it.

wenqiq commented:
It seems that the branch has conflicts with the main branch and needs to rebase.

tnqn commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,110 @@ type IPPoolList struct {
+	// The device to which the traffic should be redirected or mirrored.
+	TargetPort OVSPort `json:"targetPort"`
+
+	// The device from which the the traffic will be sent back to OVS. It should only be set for Redirect action.
```
done, thanks

tnqn commented:
/test-all

tnqn commented:
@jianjuns In latest patch I added an optional destination port parameter for VXLAN and GENEVE tunnels since some systems use 8472 for VXLAN, and changed to use each type' own term for their tunnel IDs, for example, VNI for VXLAN, Key for GRE, and SessionID of ERSPAN.

tnqn commented:
/test-all

tnqn commented on pkg/apis/crd/v1alpha2/types.go:
```diff
@@ -389,3 +389,127 @@ type IPPoolList struct {
+	// AppliedTo selects Pods to which the traffic control configuration will be applied.
+	AppliedTo AppliedTo `json:"appliedTo"`
+
+	// The direction of traffic that should be matched. It can be Ingress, Egress, or Both.
```
both ingress and egress, what would you suggest? Bidirectional？

tnqn commented:
/test-networkpolicy
